### PR TITLE
rework upgrades for malicious protocol

### PR DESCRIPTION
**Pull request by andyleiserson**
Split this off from #360.

The purpose of these changes is to allow managing total_records on the upgrade_ctx without exposing a
set_total_upgrades function on MaliciousContext.

This ended up being a bit more involved than I thought it would be originally. I could pretty easily change this to get rid of set_total_upgrades without the UpgradeContext stuff, by adding an upgrade_vector function on ContextInner and narrowing the IntoMalicious impl for IntoIter to be specific to vectors.

I do think it's possible we'll want UpgradeContext as an entity, either as part of the changes to have contexts own open channels, or as part of the changes to make context cloning more efficient (#380).

### CODE REVIEW

File: src/protocol/modulus_conversion/convert_shares.rs
Comment (benjaminsavage) [inline src/protocol/modulus_conversion/convert_shares.rs:227]: This new version is much shorter and nicer looking!

File: src/protocol/sort/generate_permutation.rs
Comment (benjaminsavage) [inline src/protocol/sort/generate_permutation.rs:273]: It's also nice you were able to take care of vectors!

File: src/protocol/boolean/generate_random_bits/malicious.rs
Comment (benjaminsavage) [inline src/protocol/boolean/generate_random_bits/malicious.rs:23]: This one seems different from all the others. Is this the only place where the client needs to provide a record_id when upgrading?

Comment (andyleiserson) [reply src/protocol/boolean/generate_random_bits/malicious.rs:23]: Yes... this is the only place where the entire input isn't presented for upgrade all at once (it's not really an input at all -- it's randomness I think used for bit decomposition), so it's not possible to refactor it such that the record IDs are generated internal to the upgrade implementation for the thing being upgraded.

It might be possible to refactor this to `upgrade_for_record<T>(record_id, input: T)` with  `T = Vector<BitConversionTriple>` (getting rid of the `_with`) if that's preferable.

File: src/protocol/context/malicious.rs
Comment (martinthomson) [inline src/protocol/context/malicious.rs:400]:
```suggestion
    for<'u> UpgradeContext<'u, F, NoRecord>: UpgradeToMalicious<T, TM> + UpgradeToMalicious<U, UM>,
```

File: src/protocol/boolean/generate_random_bits/malicious.rs
Comment (martinthomson) [reply src/protocol/boolean/generate_random_bits/malicious.rs:23]: Not really worth holding this change up.  Want to open an issue so that we can capture any ideas you have and maybe revisit later?

**Review by martinthomson — APPROVED** [review]

**Comment by akoshelev** [comment]
merging it as we have all necessary approvals

**Pull request merged**
